<?xml version="1.0"?>
<robot name = "car" xmlns:xacro="http://www.ros.org/wiki/xacro">



    <link name="base_link">
        <visual>
            <geometry>
                <!-- ox size = "1.0 0.6 0.4" -->
                <box size = "1.0 0.6 8" />
            </geometry>
            <material name = "blue">
                <color rgba = "0 0 1 1" />
            </material>
        </visual>
        <collision>
            <geometry>
                <box size= "1.0 0.6 0.4" />
            </geometry>
        </collision>
        <inertial>
            <mass value= "50.0" />
            <origin xyz= "0 0 0" rpy = "0 0 0" />
            <inertia ixx= "0.1" ixy= "0.0" ixz= "0.0" iyy= "0.1" iyz= "0.0" izz= "0.1" />
        </inertial>
    </link>

    <!-- front_left_wheel -->
    <link name="front_left_wheel">
        <visual>
            <geometry>
                <cylinder length= "0.1" radius= "0.15" />
            </geometry>
            <material name = "black">
                <color rgba = "0 0 0 1" />
            </material>
        </visual>
        <collision>
            <geometry>
                <cylinder length= "0.1" radius= "0.15" />
            </geometry>
        </collision>
        <inertial>
            <mass value= "1.0" />
            <origin xyz= "0 0 0" rpy = "0 0 0" />
            <inertia ixx= "0.01" ixy= "0.0" ixz= "0.0" iyy= "0.0" iyz= "0.0" izz= "0.01" />
        </inertial>
    </link>

    <!-- front_right_wheel -->
    <link name="front_right_wheel">
        <visual>
            <geometry>
                <cylinder length= "0.1" radius= "0.15" />
            </geometry>
            <material name = "black">
                <color rgba = "0 0 0 1" />
            </material>
        </visual>
        <collision>
            <geometry>
                <cylinder length= "0.1" radius= "0.15" />
            </geometry>
        </collision>
        <inertial>
            <mass value= "1.0" />
            <origin xyz= "0 0 0" rpy = "0 0 0" />
            <inertia ixx= "0.01" ixy= "0.0" ixz= "0.0" iyy= "0.0" iyz= "0.0" izz= "0.01" />
        </inertial>
    </link>

    <!-- rear_left_wheel -->
    <link name="rear_left_wheel">
        <visual>
            <geometry>
                <cylinder length= "0.1" radius= "0.15" />
            </geometry>

            <material name = "black">
                <color rgba = "0 0 0 1" />
            </material>
        </visual>
        <collision>
            <geometry>
                <cylinder length= "0.1" radius= "0.15" />
            </geometry>
        </collision>
        <inertial>
            <mass value= "1.0" />
            <origin xyz= "0 0 0" rpy = "0 0 0" />
            <inertia ixx= "0.01" ixy= "0.0" ixz= "0.0" iyy= "0.0" iyz= "0.0" izz= "0.01" />
        </inertial>
    </link>

    <!-- rear_right_wheel -->
    <link name="rear_right_wheel">
        <visual>
            <geometry>
                <cylinder length= "0.1" radius= "0.15" />
            </geometry>

            <material name = "black">
                <color rgba = "0 0 0 1" />
            </material>
        </visual>
        <collision>
            <geometry>
                <cylinder length= "0.1" radius= "0.15" />
            </geometry>
        </collision>
        <inertial>
            <mass value= "1.0" />
            <origin xyz= "0 0 0" rpy = "0 0 0" />
            <inertia ixx= "0.01" ixy= "0.0" ixz= "0.0" iyy= "0.0" iyz= "0.0" izz= "0.01" />
        </inertial>
    </link>

    <link name= "dummy">
    </link>

    <joint name= "dummy_joint" type= "fixed">
        <parent link= "dummy" />
        <child link= "base_link" />
    </joint>

    <!-- JOINTS -->

    <joint name="front_left_wheel_joint" type= "continuous">
        <parent link= "base_link"/>
        <child link= "front_left_wheel"/>
        <origin xyz= "0.3 0.35 -0.1" rpy= "-1.57 0 0"/>
        <axis xyz= "0 0 1"/> 

        <dynamics damping="1.0" friction="1.0"/>
    </joint>

    <joint name="front_right_wheel_joint" type= "continuous">
        <parent link= "base_link"/>
        <child link= "front_right_wheel"/>
        <origin xyz= "0.3 -0.35 -0.1" rpy= "-1.57 0 0"/>
        <axis xyz= "0 0 1"/> 

        <dynamics damping="1.0" friction="1.0"/>
    </joint>

    <joint name="rear_left_wheel_joint" type= "continuous">
        <parent link= "base_link"/>
        <child link= "rear_left_wheel"/>
        <origin xyz= "-0.3 0.35 -0.1" rpy= "-1.57 0 0"/>
        <axis xyz= "0 0 1"/> 

        <dynamics damping="1.0" friction="1.0"/>
    </joint>

    <joint name="rear_right_wheel_joint" type= "continuous">
        <parent link= "base_link"/>
        <child link= "rear_right_wheel"/>
        <origin xyz= "-0.3 -0.35 -0.1" rpy= "-1.57 0 0"/>
        <axis xyz= "0 0 1"/> 

        <dynamics damping="1.0" friction="1.0"/>
    </joint>

    <!--car controller-->
    <gazebo>
        <plugin name= "skid_steer_drive_controller" filename= "libgazebo_ros_skid_steer_drive.so">
            <updateRate>100.0</updateRate>
            <robotNamespace></robotNamespace>

            <leftFrontJoint>front_left_wheel_joint</leftFrontJoint>
            <rightFrontJoint>front_right_wheel_joint</rightFrontJoint>
            <leftRearJoint>rear_left_wheel_joint</leftRearJoint>
            <rightRearJoint>rear_right_wheel_joint</rightRearJoint>

            <wheelSeparation>0.7</wheelSeparation>
            <wheelDiameter>0.3</wheelDiameter>
            
            <torque>10</torque>

            <commandTopic>cmd_vel</commandTopic>

            <odometryTopic>odom</odometryTopic>

            <odometryFrame>odom</odometryFrame>

            <robotBaseFrame>dummy</robotBaseFrame>

            <publishWheelTF>true</publishWheelTF>

            <publishOdom>true</publishOdom>

            <publishWheelJointState>true</publishWheelJointState>

            <covariance_x>0.0001</covariance_x>
            <covariance_y>0.0001</covariance_y>
            <covariance_yaw>0.01</covariance_yaw>

        </plugin>
    </gazebo>
    


    <gazebo reference= "base_link">
        <material>Gazebo/Blue</material>
        <mu1>0.2</mu1>
        <mu2>0.2</mu2>
    </gazebo>

    <gazebo reference= "front_left_wheel">
        <material>Gazebo/Black</material>
        <mu1>0.2</mu1>
        <mu2>0.2</mu2>
    </gazebo>
    <gazebo reference= "front_right_wheel">
        <material>Gazebo/Black</material>
        <mu1>0.2</mu1>
        <mu2>0.2</mu2>
    </gazebo>
    <gazebo reference= "rear_left_wheel">
        <material>Gazebo/Black</material>
        <mu1>0.2</mu1>
        <mu2>0.2</mu2>
    </gazebo>
    <gazebo reference= "rear_right_wheel">
        <material>Gazebo/Black</material>
        <mu1>0.2</mu1>
        <mu2>0.2</mu2>
    </gazebo>

    <gazebo reference= "camera_link">
        <material>Gazebo/Red</material>
        
    </gazebo>


    <!-- depth camera imitation of sonar, taken from the box.urdf code -->
    <link name= "depth_camera_link">
        <visual>
            <geometry> 
                <box size= "0.1 0.1 0.1" />
            </geometry>
            <material name= "red" >
                <color rgba= "1 0 0 1" />
            </material>
        </visual>
        <inertial>
            <mass value= "0.2" />
            <origin xyz= "0 0 0" />
            <inertia ixx= "0.1" ixy= "0.0" ixz= "0.0" iyy= "0.1" iyz= "0.0" izz= "0.1" />
        </inertial>
        <collision>
            <geometry>
                <box size= "0.1 0.1 0.1" />
            </geometry>
        </collision>
    </link>

    <joint name= "depth_camera_joint" type= "fixed">
        <parent link= "base_link" />
        <child link= "depth_camera_link" />
        <origin xyz= "0.55 0 3 " rpy= "0 0 0"/>
    </joint> 

    <!-- optical link and joint for rviz -->
    <link name= "depth_camera_optical_link"></link>

    <joint name= "depth_camera_optical_joint" type= "fixed">
        <parent link= "depth_camera_link" />
        <child link= "depth_camera_optical_link" />
        <origin xyz= "0 0 0 " rpy= "-1.57 0 -1.57"/>
    </joint>

    <gazebo reference= "depth_camera_link">
        <material>Gazebo/Red</material>
        <sensor name= "depth_camera_link" type= "depth">
            <visualize>true</visualize>
            <updateRate>10</updateRate>
            <camera>
                <horizontal_fov>0.785</horizontal_fov> 
                <vertical_fov>0.785</vertical_fov>   
                <image>
                    <width>10</width>
                    <height>10</height>
                </image>
                <clip>
                    <near>0.1</near>
                    <far>5.0</far>
                </clip>
            </camera>
            <plugin name="depth_camera_sensor" filename="libgazebo_ros_openni_kinect.so">
                <cameraName>depth_camera</cameraName>
                <frameName>depth_camera_optical_link</frameName>
                <alwaysOn>true</alwaysOn>
                <pointCloudTopicName>points</pointCloudTopicName>
                
                <imageTopicName>rgb/image_raw</imageTopicName>
                <depthImageTopicName>depth/image_raw</depthImageTopicName>
                <cameraInfoTopicName>rgb/camera_info</cameraInfoTopicName>
                <depthImageCameraInfoTopicName>depth/camera_info</depthImageCameraInfoTopicName>
                
                <!-- <baseline>0.1</baseline> -->
                <!-- <hfov>1.0472</hfov> 
                <vfov>1.0472</vfov>  -->
            </plugin>
        </sensor>
    </gazebo>

    </robot>

    
    <!-- <gazebo>
    <plugin name="differential_drive_controller" filename="libgazebo_ros_diff_drive.so">
        <commandTopic>cmd_vel</commandTopic>
        <odometryTopic>odom</odometryTopic>
        <odometryFrame>odom</odometryFrame>
        <robotBaseFrame>base_link</robotBaseFrame>
        <publishWheelTF>true</publishWheelTF>
        <publishWheelJointState>true</publishWheelJointState>
        <wheelSeparation>0.7</wheelSeparation> 
        <wheelDiameter>0.3</wheelDiameter> 
        <torque>10.0</torque> 
        <updateRate>10.0</updateRate>

        <leftJoint>front_left_wheel_joint</leftJoint>
        <rightJoint>front_right_wheel_joint</rightJoint>
    </plugin>
    </gazebo> -->

<!-- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -->
  
    <!-- camera -->
    <!-- <link name= "camera_link">
        <visual>
            <geometry> 
                <box size= "0.1 0.1 0.1" />
            </geometry>
            <material name= "red" >
                <color rgba= "1 0 0 1" />
            </material>
        </visual>
        <inertial>
            <mass value= "0.2" />
            <origin xyz= "0 0 0" />
            <inertia ixx= "0.1" ixy= "0.0" ixz= "0.0" iyy= "0.1" iyz= "0.0" izz= "0.1" />
        </inertial>
        <collision>
            <geometry>
                <box size= "0.1 0.1 0.1" />
            </geometry>
        </collision>
    </link>


    <joint name= "camera_joint" type= "fixed">
        <parent link= "base_link" />
        <child link= "camera_link" />
        <origin xyz= "0.4 0 0.25 " rpy= "0 0 0"/>
    </joint> 

    <link name= "camera_link_optical"></link>

    <joint name= "camera_optical_joint" type= "fixed">
        <parent link= "camera_link" />
        <child link= "camera_link_optical" />
        <origin xyz= "0 0 0 " rpy= "-1.57 0 -1.57"/>
    </joint> 

    <gazebo reference= "camera_link">
        <material>Gazebo/Red</material>
        <sensor type= "camera" name= "camera_sensor">
            <pose>0 0 0 0 0 0</pose>
            <visualize>false</visualize>
            <update_rate>30</update_rate>
            <camera>
                <horizontal_fov>1.047</horizontal_fov>
                <image>
                    <width>640</width>
                    <height>480</height>
                    <format>R8G8B8</format>
                </image>
                <clip>
                    <near>0.1</near>
                    <far>10</far>
                </clip>
            </camera>
            <plugin name= "camera_controller" filename= "libgazebo_ros_camera.so" >
                <frameName>camera_link_optical</frameName>
                <alwaysOn>true</alwaysOn>
                <updateRate>0.0</updateRate>
                <cameraName>camera</cameraName>
                <imageTopicName>image_raw</imageTopicName>
                <cameraInfoTopicName>camera_info</cameraInfoTopicName>
            </plugin>
        </sensor>
    </gazebo> -->

